django_ratelimit/core.py

```diff
@@ -227,45 +248,96 @@ def get_usage(request, group=None, fn=None, key=None, rate=None, method=ALL,
     cache_key = _make_cache_key(group, window, rate, value, method)
 
     count = None
-    try:
-        added = cache.add(cache_key, initial_value, period + EXPIRATION_FUDGE)
-    except socket.gaierror:  # for redis
-        added = False
-    if added:
-        count = initial_value
+    if is_async:
+        async def inner():
+            try:
+                # Some caches don't have an async implementation
+                if getattr(cache, 'aadd') is not None:
+                    added = await cache.aadd(cache_key, initial_value, period + EXPIRATION_FUDGE)
+                else:
+                    added = await cache.add(cache_key, initial_value, period + EXPIRATION_FUDGE)
+            except socket.gaierror:  # for redis
+                added = False
+            if added:
+                count = initial_value
+            else:
+                if increment:
+                    try:
+                        # python3-memcached will throw a ValueError if the server is
+                        # unavailable or (somehow) the key doesn't exist. redis, on the
+                        # other hand, simply returns None.
+                        count = cache.incr(cache_key)
```

Comment:
Looks like aincr and adecr have been around since ~4.0 but [the implementation isn't atomic](https://github.com/django/django/blame/main/django/core/cache/backends/base.py#L274-L281), even in caches that should support atomic incr/decr. That's going to be a challenge, even if it gets fixed. I think this is the best choice for now, even if it

Reply:
if they are around now then i could add the if statement to the call as well. That will let us support them if the implementation becomes atomic in the future.